src/easy.rs

```diff
@@ -0,0 +1,439 @@
+/// serialized into bytes to get passed to the underlying byte stream. Each
+/// frame written to `EasyFramed` will be serialized with this trait to an internal
+/// buffer. That buffer is then written out when possible to the underlying I/O
+/// stream.
+pub trait Serialize {
+
+    /// The frame that's being serialized to a byte buffer.
+    ///
+    /// This type is the type of frame that's also being written to a `EasyFramed`.
+    type In;
+
+    /// Serializes a frame into the buffer provided.
+    ///
+    /// This method will serialize `msg` into the byte buffer provided by `buf`.
+    /// The `buf` provided is an internal buffer of the `EasyFramed` instance and
+    /// will be written out when possible.
+    fn serialize(&mut self, msg: Self::In, buf: &mut Vec<u8>);
+}
+
+impl<T, P, S> EasyFramed<T, P, S>
+    where T: Io,
+          P: Parse,
+          S: Serialize,
+{
+    /// Creates a new instance of `EasyFramed` from the given component pieces.
+    ///
+    /// This method will create a new instance of `EasyFramed` which implements
+    /// `FramedIo` for reading and writing frames from an underlying I/O stream.
+    /// The `upstream` argument here is the byte-based I/O stream that it will
+    /// be operating on. Data will be read from this stream and parsed with
+    /// `parse` into frames. Frames written to this instance will be serialized
+    /// by `serialize` and then written to `upstream`.
+    ///
+    /// The `rd` and `wr` buffers provided are used for reading and writing
+    /// bytes and provide a small amount of control over how buffering happens.
+    pub fn new(upstream: T,
+               parse: P,
+               serialize: S) -> EasyFramed<T, P, S> {
+
+        trace!("creating new framed transport");
+        EasyFramed {
+            upstream: upstream,
+            parse: parse,
+            serialize: serialize,
+            is_readable: false,
+            eof: false,
+            rd: EasyBuf::new(),
+            wr: Vec::with_capacity(8 * 1024),
+        }
+    }
+}
+
+impl<T, P, S> FramedIo for EasyFramed<T, P, S>
+    where T: Io,
+          P: Parse,
+          S: Serialize,
+{
+    type In = S::In;
+    type Out = P::Out;
+
+    fn poll_read(&mut self) -> Async<()> {
+        if self.is_readable || self.upstream.poll_read().is_ready() {
+            Async::Ready(())
+        } else {
+            Async::NotReady
+        }
+    }
+
+    fn read(&mut self) -> Poll<Self::Out, io::Error> {
+        loop {
+            // If the read buffer has any pending data, then it could be
+            // possible that `parse` will return a new frame. We leave it to
+            // the parser to optimize detecting that more data is required.
+            if self.is_readable {
+                if self.eof {
+                    if self.rd.len() == 0 {
+                        // TODO: seems like this should return `None`?
+                        return Ok(Async::NotReady)
```

Comment:
Interesting! This made me want to change the trait to return `Poll<Option<Frame>, ...>` so signal EOF. What protocols though have a frame at the end of a stream when there are no bytes available?